tempesta_fw/http_sess.c

```diff
@@ -402,51 +438,246 @@ tfw_http_sticky_add(TfwHttpResp *resp)
 	return r;
 }
 
+/*
+ * Calculate HMAC value for redirection mark.
+ *
+ * HMAC value is based on:
+ * - Number of attempts for session establishing;
+ * - Timestamp of first attempt;
+ * - Secret key.
+ */
+static int
+__redir_hmac_calc(TfwHttpReq *req, RedirMarkVal *mv)
+{
+	int r;
+	char desc[sizeof(struct shash_desc)
+		  + crypto_shash_descsize(tfw_sticky_shash)]
+		  CRYPTO_MINALIGN_ATTR;
+	struct shash_desc *shash_desc = (struct shash_desc *)desc;
+
+	bzero_fast(desc, sizeof(desc));
+	shash_desc->tfm = tfw_sticky_shash;
+	shash_desc->flags = 0;
+
+	TFW_DBG("http_sess: calculate redirection mark: ts=%#lx(now=%#lx),"
+		" att_no=%#x\n", mv->ts, jiffies, mv->att_no);
+
+	if ((r = crypto_shash_init(shash_desc)))
+		return r;
+	r = crypto_shash_update(shash_desc, (u8 *)&mv->att_no, sizeof(mv->att_no));
+	if (r)
+		return r;
+	return crypto_shash_finup(shash_desc, (u8 *)&mv->ts, sizeof(mv->ts),
+				  mv->hmac);
+}
+
+static int
+tfw_http_redir_mark_get(TfwHttpReq *req, TfwStr *out_val)
+{
+	TfwStr *mark = &req->mark;
+	TfwStr *c, *end;
+
+	if (TFW_STR_EMPTY(mark))
+		return 0;
+
+	/* Find the value chunk. */
+	end = (TfwStr*)mark->ptr + TFW_STR_CHUNKN(mark);
+	for (c = mark->ptr; c != end; ++c)
+		if (c->flags & TFW_STR_VALUE)
+			break;
+
+	BUG_ON(c == end);
+
+	tfw_str_collect_cmp(c, end, out_val, NULL);
+
+	return 1;
+}
+
+#define sess_warn(check, addr, fmt, ...)				\
+do {									\
+	char abuf[TFW_ADDR_STR_BUF_SIZE] = {0};				\
+	tfw_addr_fmt_v6(&(addr)->v6.sin6_addr, 0, abuf);		\
+	TFW_WARN("http_sess: %s for %s" fmt, check, abuf, ##__VA_ARGS__); \
+} while (0)
+
+/* The set of macros for parsing hex strings of following format:
+ *
+ *    <value_1> | <value_2> | ... | <value_n> | HMAC_value
+ *
+ * and HMAC value verification. In 'HEX_STR_TO_BIN_HMAC()' macro the
+ * 'hmac' parameter must contain newly recalculated HMAC value.
+ */
+#define HEX_STR_TO_BIN_INIT(tr, c, str, end)				\
+	tr = NULL;							\
+	TFW_STR_FOR_EACH_CHUNK_INIT(c, str, end);
+
+#define HEX_STR_TO_BIN_GET(obj, f)					\
+({									\
+	int count = 0;							\
+	if (!tr && c < end)						\
+		tr = c->ptr;						\
+	for ( ; c < end; ++c) {						\
+		for ( ; tr < (unsigned char *)c->ptr + c->len; ++tr) {	\
+			if (count++ == sizeof((obj)->f) * 2)		\
+				goto end_##f;				\
+			(obj)->f = ((obj)->f << 4) + hex_to_bin(*tr);	\
+		}							\
+	}								\
+end_##f:								\
+	count;								\
```

Comment:
It seems the result value of the statement is never used

Reply:
Yes, but `gcc` does not allow labels at the end of compound statement.

Reply:
Can you then just use empty statement `;` instead?

Reply:
Yes, missed that - thanks.
Corrected.